jobserver/models/job_request.py

```diff
@@ -43,17 +43,17 @@ def with_started_at(self):
 class JobRequestManager(models.Manager.from_queryset(JobRequestQuerySet)):
     use_in_migrations = True
 
-    def previous(self, job_request):
-        return (
-            super()
-            .filter(
-                workspace=job_request.workspace,
-                backend=job_request.backend,
-                id__lt=job_request.id,
-            )
-            .order_by("created_at")
-            .last()
+    def previous(self, job_request, succeeded=None):
+        workspace_backend_job_requests = super().filter(
+            workspace=job_request.workspace,
+            backend=job_request.backend,
+            id__lt=job_request.id,
         )
+        if succeeded:
+            workspace_backend_job_requests = workspace_backend_job_requests.annotate(
+                min_status=Min("jobs__status")
+            ).filter(min_status="succeeded")
```

Comment:
You're going to have to explain to me what you're doing with `Min` here, because I've read it, run it in the shell, and I can't shake the feeling that it's a hack.

Reply:
the values for `job.status` are ("failed", "pending", "running", "succeeded"). An efficient way of answering the question: "are all of the status values for all of this `JobRequest`'s `Job`s equal to 'succeeded'?" is to take the minimum (by alphabetical sort) of the `Job`s' `status`es. Which is to say: `min("failed", "pending", "running", "succeeded") == "succeeded"`

This could be made more robust by adding a `CheckConstraint` to that field to enforce the fact those are indeed the only permissible values or by 

```python
workspace_backend_job_requests = workspace_backend_job_requests.annotate(
                min_status=Min("jobs__status"), max_status=Max("jobs__status")            
               ).filter(min_status="succeeded", max_status="succeeded")
```

such that if "zzz" was added as a status then `(min("succeeded","zzz")=="succeeded")&(max("succeeded","zzz")=="succeeded")==False`


To use the logic in the `@property` of `Job.status` involves running the query to get all the `JobRequests` prior to this one for this workspace/backend, then filtering in memory for the "succeeded" ones. By using the aggregate(s) above, this filtering can be pushed into the database query.

Reply:
see https://github.com/opensafely-core/job-server/pull/4130/commits/35ccbf88e41d832ac13566c7269675e8ef952ed4